File: packages/sdk-typescript/src/services/user-service.ts
/**
 * User Service - User operations with verification.
 */

import type { ISLClientConfig } from '../config';
import type {
  User,
  CreateUserInput,
  UpdateUserInput,
  ListUsersInput,
  SearchUsersInput,
} from '../models';
import { parseUser } from '../models';
import type {
  CreateUserResult,
  GetUserResult,
  UpdateUserResult,
  DeleteUserResult,
  ListUsersResult,
  SearchUsersResult,
} from '../results';
import { ok, err, CreateUserErrors, GetUserErrors } from '../results';
import { RuntimeChecker } from '../verification';

/**
 * User service for user-related operations
 */
export class UserService {
  private readonly config: Required<ISLClientConfig>;
  private readonly checker: RuntimeChecker;
  private readonly fetchFn: typeof fetch;

  constructor(config: Required<ISLClientConfig>) {
    this.config = config;
    this.checker = new RuntimeChecker(config.verification);
    this.fetchFn = config.fetch ?? globalThis.fetch;
  }

  /**
   * Create a new user
   */
  async createUser(input: CreateUserInput): Promise<CreateUserResult> {
    // Verify preconditions
    if (this.config.verification?.enablePreconditions) {
      this.checker.verifyCreateUserPreconditions(input);
    }

    try {
      const response = await this.request('POST', '/api/users', input);

      if (response.ok) {
        const user = parseUser(await response.json());

        // Verify postconditions
        if (this.config.verification?.enablePostconditions) {
          this.checker.verifyCreateUserPostconditions(input, user);
        }

        return ok(user);
      }

      return this.handleCreateUserError(response);
    } catch (error) {
      return err(CreateUserErrors.networkError(error as Error));
    }
  }

  /**
   * Get a user by ID
   */
  async getUser(userId: string): Promise<GetUserResult> {
    // Verify preconditions
    if (this.config.verification?.enablePreconditions) {
      this.checker.verifyGetUserPreconditions(userId);
    }

    try {
      const response = await this.request('GET', `/api/users/${userId}`);

      if (response.ok) {
        const user = parseUser(await response.json());

        // Verify postconditions
        if (this.config.verification?.enablePostconditions) {
          this.checker.verifyGetUserPostconditions(userId, user);
        }

        return ok(user);
      }

      if (response.status === 404) {
        return err(GetUserErrors.notFound());
      }

      if (response.status === 401) {
        return err(GetUserErrors.unauthorized());
      }

      return err(GetUserErrors.serverError(`Server error: ${response.status}`));
    } catch (error) {
      return err(GetUserErrors.networkError(error as Error));
    }
  }

  /**
   * Update a user
   */
  async updateUser(userId: string, input: UpdateUserInput): Promise<UpdateUserResult> {
    // Verify preconditions
    if (this.config.verification?.enablePreconditions) {
      this.checker.verifyUpdateUserPreconditions(userId, input);
    }

    try {
      const response = await this.request('PATCH', `/api/users/${userId}`, input);

      if (response.ok) {
        const user = parseUser(await response.json());

        // Verify postconditions
        if (this.config.verification?.enablePostconditions) {
          this.checker.verifyUpdateUserPostconditions(userId, input, user);
        }

        return ok(user);
      }

      return this.handleUpdateUserError(response);
    } catch (error) {
      return err({ code: 'NETWORK_ERROR', cause: error as Error });
    }
  }

  /**
   * Delete a user
   */
  async deleteUser(userId: string): Promise<DeleteUserResult> {
    // Verify preconditions
    if (this.config.verification?.enablePreconditions) {
      this.checker.verifyDeleteUserPreconditions(userId);
    }

    try {
      const response = await this.request('DELETE', `/api/users/${userId}`);

      if (response.ok || response.status === 204) {
        return ok(undefined);
      }

      if (response.status === 404) {
        return err({ code: 'NOT_FOUND' });
      }

      if (response.status === 401) {
        return err({ code: 'UNAUTHORIZED' });
      }

      if (response.status === 403) {
        return err({ code: 'FORBIDDEN' });
      }

      return err({ code: 'SERVER_ERROR', message: `Server error: ${response.status}` });
    } catch (error) {
      return err({ code: 'NETWORK_ERROR', cause: error as Error });
    }
  }

  /**
   * List users with pagination
   */
  async listUsers(input: ListUsersInput = {}): Promise<ListUsersResult> {
    // Verify preconditions
    if (this.config.verification?.enablePreconditions) {
      this.checker.verifyListUsersPreconditions(input);
    }

    try {
      const params = new URLSearchParams();
      if (input.status) params.set('status', input.status);
      if (input.role) params.set('role', input.role);
      if (input.pageSize) params.set('page_size', String(input.pageSize));
      if (input.pageToken) params.set('page_token', input.pageToken);
      if (input.sortBy) params.set('sort_by', input.sortBy);
      if (input.sortOrder) params.set('sort_order', input.sortOrder);

      const url = `/api/users${params.toString() ? `?${params}` : ''}`;
      const response = await this.request('GET', url);

      if (response.ok) {
        const data = await response.json();
        const users = (data.items || []).map(parseUser);
        return ok({
          users,
          nextPageToken: data.next_page_token,
          totalCount: data.total_count,
        });
      }

      if (response.status === 401) {
        return err({ code: 'UNAUTHORIZED' });
      }

      return err({ code: 'SERVER_ERROR', message: `Server error: ${response.status}` });
    } catch (error) {
      return err({ code: 'NETWORK_ERROR', cause: error as Error });
    }
  }

  /**
   * Search users by query
   */
  async searchUsers(input: SearchUsersInput): Promise<SearchUsersResult> {
    // Verify preconditions
    if (this.config.verification?.enablePreconditions) {
      this.checker.verifySearchUsersPreconditions(input);
    }

    try {
      const params = new URLSearchParams();
      params.set('q', input.query);
      if (input.fields) params.set('fields', input.fields.join(','));
      if (input.pageSize) params.set('page_size', String(input.pageSize));
      if (input.pageToken) params.set('page_token', input.pageToken);

      const url = `/api/users/search?${params}`;
      const response = await this.request('GET', url);

      if (response.ok) {
        const data = await response.json();
        const users = (data.items || []).map(parseUser);
        return ok({
          users,
          nextPageToken: data.next_page_token,
          totalCount: data.total_count,
        });
      }

      if (response.status === 401) {
        return err({ code: 'UNAUTHORIZED' });
      }

      return err({ code: 'SERVER_ERROR', message: `Server error: ${response.status}` });
    } catch (error) {
      return err({ code: 'NETWORK_ERROR', cause: error as Error });
    }
  }

  /**
   * Observe user updates via WebSocket
   */
  async *observe(userId: string): AsyncIterableIterator<User> {
    const wsUrl = this.config.baseUrl.replace(/^http/, 'ws');
    const ws = new WebSocket(`${wsUrl}/ws/users/${userId}`);

    try {
      while (true) {
        const event = await new Promise<MessageEvent>((resolve, reject) => {
          ws.onmessage = resolve;
          ws.onerror = reject;
          ws.onclose = () => reject(new Error('WebSocket closed'));
        });

        yield parseUser(JSON.parse(event.data));
      }
    } finally {
      ws.close();
    }
  }

  // ===========================================================================
  // Private Methods
  // ===========================================================================

  private async request(
    method: string,
    path: string,
    body?: unknown
  ): Promise<Response> {
    const url = `${this.config.baseUrl}${path}`;

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      'X-Client-SDK': 'isl-typescript/0.1.0',
      ...this.config.headers,
    };

    if (this.config.authToken) {
      headers['Authorization'] = `Bearer ${this.config.authToken}`;
    }

    let request = new Request(url, {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
    });

    // Apply request interceptor
    if (this.config.interceptors?.request) {
      request = await this.config.interceptors.request(request);
    }

    let response = await this.fetchFn(request);

    // Apply response interceptor
    if (this.config.interceptors?.response) {
      response = await this.config.interceptors.response(response);
    }

    return response;
  }

  private async handleCreateUserError(response: Response): Promise<CreateUserResult> {
    if (response.status === 409) {
      const body = await response.json().catch(() => ({}));
      const message = (body.message || '').toLowerCase();
      if (message.includes('email')) {
        return err(CreateUserErrors.duplicateEmail());
      }
      if (message.includes('username')) {
        return err(CreateUserErrors.duplicateUsername());
      }
      return err(CreateUserErrors.invalidInput(body.message || 'Conflict'));
    }

    if (response.status === 400) {
      const body = await response.json().catch(() => ({}));
      return err(CreateUserErrors.invalidInput(body.message || 'Invalid input', body.field));
    }

    if (response.status === 429) {
      const retryAfter = parseInt(response.headers.get('Retry-After') || '60', 10);
      return err(CreateUserErrors.rateLimited(retryAfter));
    }

    return err(CreateUserErrors.serverError(`Server error: ${response.status}`));
  }

  private async handleUpdateUserError(response: Response): Promise<UpdateUserResult> {
    if (response.status === 404) {
      return err({ code: 'NOT_FOUND' });
    }

    if (response.status === 401) {
      return err({ code: 'UNAUTHORIZED' });
    }

    if (response.status === 403) {
      return err({ code: 'FORBIDDEN' });
    }

    if (response.status === 400) {
      const body = await response.json().catch(() => ({}));
      return err({ code: 'INVALID_INPUT', message: body.message || 'Invalid input', field: body.field });
    }

    if (response.status === 409) {
      const body = await response.json().catch(() => ({}));
      return err({ code: 'CONFLICT', message: body.message || 'Conflict' });
    }

    if (response.status === 429) {
      const retryAfter = parseInt(response.headers.get('Retry-After') || '60', 10);
      return err({ code: 'RATE_LIMITED', retryAfter });
    }

    return err({ code: 'SERVER_ERROR', message: `Server error: ${response.status}` });
  }
}

File: packages/sdk-kotlin/src/main/kotlin/com/isl/client/services/UserService.kt
package com.isl.client.services

import com.isl.client.models.*
import com.isl.client.networking.ApiClient
import com.isl.client.networking.ApiResponse
import com.isl.client.validation.Validators
import com.isl.client.verification.RuntimeCheck
import com.isl.client.verification.VerificationConfig
import com.isl.client.verification.Verified
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.channels.awaitClose

/**
 * User service for user-related operations
 * 
 * This service implements all user behaviors defined in the ISL specification
 * with automatic precondition and postcondition verification.
 */
class UserService(
    private val apiClient: ApiClient,
    private val verificationConfig: VerificationConfig = VerificationConfig()
) {
    
    private val runtimeCheck = RuntimeCheck(verificationConfig)
    
    // ========================================================================
    // Create User
    // ========================================================================
    
    /**
     * Create a new user
     * 
     * Preconditions:
     * - Email must be valid format (contains @, max 254 chars)
     * - Username must be 3-30 characters
     * 
     * Postconditions:
     * - User created with PENDING status
     * - User email matches input
     * - User username matches input
     * 
     * @param input The user creation input
     * @return CreateUserResult with success or typed error
     */
    @Verified(
        preconditions = [
            "input.email.contains('@')",
            "input.email.length <= 254",
            "input.username.length in 3..30"
        ],
        postconditions = [
            "result.email == input.email",
            "result.username == input.username",
            "result.status == UserStatus.PENDING"
        ]
    )
    suspend fun createUser(input: CreateUserInput): CreateUserResult {
        // Validate preconditions
        if (verificationConfig.enablePreconditions) {
            val validationResult = Validators.validateCreateUserInput(input)
            if (validationResult is com.isl.client.validation.ValidationResult.Invalid) {
                return CreateUserResult.Error.InvalidInput(
                    message = validationResult.errors.joinToString("; "),
                    field = validationResult.field
                )
            }
            runtimeCheck.verifyCreateUserPreconditions(input)
        }
        
        // Make API request
        val response = apiClient.post<User, CreateUserInput>("/api/users", input)
        
        return when (response) {
            is ApiResponse.Success -> {
                val user = response.data
                
                // Verify postconditions
                if (verificationConfig.enablePostconditions) {
                    runtimeCheck.verifyCreateUserPostconditions(input, user)
                }
                
                CreateUserResult.Success(user)
            }
            is ApiResponse.Error -> {
                mapCreateUserError(response)
            }
            is ApiResponse.RateLimited -> {
                CreateUserResult.Error.RateLimited(response.retryAfterSeconds)
            }
        }
    }
    
    // ========================================================================
    // Get User
    // ========================================================================
    
    /**
     * Get a user by ID
     * 
     * @param userId The user ID to fetch
     * @return GetUserResult with success or typed error
     */
    @Verified(
        preconditions = ["userId.isNotBlank()"],
        postconditions = ["result.id == userId"]
    )
    suspend fun getUser(userId: String): GetUserResult {
        // Validate preconditions
        if (verificationConfig.enablePreconditions) {
            runtimeCheck.verifyGetUserPreconditions(userId)
        }
        
        val response = apiClient.get<User>("/api/users/$userId")
        
        return when (response) {
            is ApiResponse.Success -> {
                val user = response.data
                
                // Verify postconditions
                if (verificationConfig.enablePostconditions) {
                    runtimeCheck.verifyGetUserPostconditions(userId, user)
                }
                
                GetUserResult.Success(user)
            }
            is ApiResponse.Error -> {
                when (response.errorCode) {
                    ErrorCodes.NOT_FOUND -> GetUserResult.Error.NotFound
                    ErrorCodes.UNAUTHORIZED -> GetUserResult.Error.Unauthorized
                    else -> GetUserResult.Error.ServerError(response.message)
                }
            }
            is ApiResponse.RateLimited -> {
                GetUserResult.Error.ServerError("Rate limited")
            }
        }
    }
    
    // ========================================================================
    // Update User
    // ========================================================================
    
    /**
     * Update a user
     * 
     * @param userId The user ID to update
     * @param input The update input
     * @return UpdateUserResult with success or typed error
     */
    @Verified(
        preconditions = [
            "userId.isNotBlank()",
            "input.username?.length in 3..30 || input.username == null"
        ],
        postconditions = [
            "result.id == userId",
            "result.username == input.username || input.username == null"
        ]
    )
    suspend fun updateUser(userId: String, input: UpdateUserInput): UpdateUserResult {
        // Validate preconditions
        if (verificationConfig.enablePreconditions) {
            val validationResult = Validators.validateUpdateUserInput(input)
            if (validationResult is com.isl.client.validation.ValidationResult.Invalid) {
                return UpdateUserResult.Error.InvalidInput(
                    message = validationResult.errors.joinToString("; "),
                    field = validationResult.field
                )
            }
            runtimeCheck.verifyUpdateUserPreconditions(userId, input)
        }
        
        val response = apiClient.patch<User, UpdateUserInput>("/api/users/$userId", input)
        
        return when (response) {
            is ApiResponse.Success -> {
                val user = response.data
                
                // Verify postconditions
                if (verificationConfig.enablePostconditions) {
                    runtimeCheck.verifyUpdateUserPostconditions(userId, input, user)
                }
                
                UpdateUserResult.Success(user)
            }
            is ApiResponse.Error -> {
                when (response.errorCode) {
                    ErrorCodes.NOT_FOUND -> UpdateUserResult.Error.NotFound
                    ErrorCodes.UNAUTHORIZED -> UpdateUserResult.Error.Unauthorized
                    ErrorCodes.FORBIDDEN -> UpdateUserResult.Error.Forbidden
                    ErrorCodes.CONFLICT -> UpdateUserResult.Error.Conflict(response.message)
                    ErrorCodes.INVALID_INPUT -> UpdateUserResult.Error.InvalidInput(response.message)
                    else -> UpdateUserResult.Error.ServerError(response.message)
                }
            }
            is ApiResponse.RateLimited -> {
                UpdateUserResult.Error.RateLimited(response.retryAfterSeconds)
            }
        }
    }
    
    // ========================================================================
    // Delete User
    // ========================================================================
    
    /**
     * Delete a user
     * 
     * @param userId The user ID to delete
     * @return DeleteUserResult with success or typed error
     */
    @Verified(
        preconditions = ["userId.isNotBlank()"]
    )
    suspend fun deleteUser(userId: String): DeleteUserResult {
        // Validate preconditions
        if (verificationConfig.enablePreconditions) {
            runtimeCheck.verifyDeleteUserPreconditions(userId)
        }
        
        val response = apiClient.delete<Unit>("/api/users/$userId")
        
        return when (response) {
            is ApiResponse.Success -> DeleteUserResult.Success
            is ApiResponse.Error -> {
                when (response.errorCode) {
                    ErrorCodes.NOT_FOUND -> DeleteUserResult.Error.NotFound
                    ErrorCodes.UNAUTHORIZED -> DeleteUserResult.Error.Unauthorized
                    ErrorCodes.FORBIDDEN -> DeleteUserResult.Error.Forbidden
                    else -> DeleteUserResult.Error.ServerError(response.message)
                }
            }
            is ApiResponse.RateLimited -> {
                DeleteUserResult.Error.ServerError("Rate limited")
            }
        }
    }
    
    // ========================================================================
    // List Users
    // ========================================================================
    
    /**
     * List users with pagination
     * 
     * @param input The list input with filters and pagination
     * @return ListUsersResult with success or typed error
     */
    suspend fun listUsers(input: ListUsersInput = ListUsersInput()): ListUsersResult {
        // Validate preconditions
        if (verificationConfig.enablePreconditions) {
            val validationResult = Validators.validateListUsersInput(input)
            if (validationResult is com.isl.client.validation.ValidationResult.Invalid) {
                return ListUsersResult.Error.InvalidInput(validationResult.errors.joinToString("; "))
            }
        }
        
        val queryParams = buildMap {
            input.status?.let { put("status", it.name) }
            input.role?.let { put("role", it.name) }
            put("pageSize", input.pageSize.toString())
            input.pageToken?.let { put("pageToken", it) }
            put("sortBy", input.sortBy)
            put("sortOrder", input.sortOrder.name)
        }
        
        val response = apiClient.get<PaginatedList<User>>("/api/users", queryParams)
        
        return when (response) {
            is ApiResponse.Success -> {
                val data = response.data
                ListUsersResult.Success(
                    users = data.items,
                    nextPageToken = data.nextPageToken,
                    totalCount = data.totalCount
                )
            }
            is ApiResponse.Error -> {
                when (response.errorCode) {
                    ErrorCodes.UNAUTHORIZED -> ListUsersResult.Error.Unauthorized
                    ErrorCodes.INVALID_INPUT -> ListUsersResult.Error.InvalidInput(response.message)
                    else -> ListUsersResult.Error.ServerError(response.message)
                }
            }
            is ApiResponse.RateLimited -> {
                ListUsersResult.Error.ServerError("Rate limited")
            }
        }
    }
    
    // ========================================================================
    // Search Users
    // ========================================================================
    
    /**
     * Search users by query
     * 
     * @param input The search input
     * @return SearchUsersResult with success or typed error
     */
    suspend fun searchUsers(input: SearchUsersInput): SearchUsersResult {
        // Validate preconditions
        if (verificationConfig.enablePreconditions) {
            val validationResult = Validators.validateSearchUsersInput(input)
            if (validationResult is com.isl.client.validation.ValidationResult.Invalid) {
                return SearchUsersResult.Error.InvalidInput(validationResult.errors.joinToString("; "))
            }
        }
        
        val queryParams = buildMap {
            put("q", input.query)
            put("fields", input.fields.joinToString(","))
            put("pageSize", input.pageSize.toString())
            input.pageToken?.let { put("pageToken", it) }
        }
        
        val response = apiClient.get<PaginatedList<User>>("/api/users/search", queryParams)
        
        return when (response) {
            is ApiResponse.Success -> {
                val data = response.data
                SearchUsersResult.Success(
                    users = data.items,
                    nextPageToken = data.nextPageToken,
                    totalCount = data.totalCount
                )
            }
            is ApiResponse.Error -> {
                when (response.errorCode) {
                    ErrorCodes.UNAUTHORIZED -> SearchUsersResult.Error.Unauthorized
                    ErrorCodes.INVALID_INPUT -> SearchUsersResult.Error.InvalidInput(response.message)
                    else -> SearchUsersResult.Error.ServerError(response.message)
                }
            }
            is ApiResponse.RateLimited -> {
                SearchUsersResult.Error.ServerError("Rate limited")
            }
        }
    }
    
    // ========================================================================
    // Real-time Updates
    // ========================================================================
    
    /**
     * Observe user updates via WebSocket
     * 
     * @param userId The user ID to observe
     * @return Flow of User updates
     */
    fun observeUser(userId: String): Flow<User> {
        if (verificationConfig.enablePreconditions) {
            require(userId.isNotBlank()) { "User ID cannot be blank" }
        }
        
        return apiClient.webSocket<UserUpdateEvent>("/ws/users/$userId") { event ->
            // Event handler
        }.map { event ->
            event.user
        }
    }
    
    /**
     * Observe all user updates via WebSocket
     * 
     * @return Flow of UserUpdateEvent
     */
    fun observeAllUsers(): Flow<UserUpdateEvent> {
        return apiClient.webSocket("/ws/users") { }
    }
    
    // ========================================================================
    // Error Mapping
    // ========================================================================
    
    private fun mapCreateUserError(response: ApiResponse.Error): CreateUserResult.Error {
        return when (response.errorCode) {
            ErrorCodes.DUPLICATE_EMAIL -> CreateUserResult.Error.DuplicateEmail
            ErrorCodes.DUPLICATE_USERNAME -> CreateUserResult.Error.DuplicateUsername
            ErrorCodes.CONFLICT -> {
                if (response.message.contains("email", ignoreCase = true)) {
                    CreateUserResult.Error.DuplicateEmail
                } else if (response.message.contains("username", ignoreCase = true)) {
                    CreateUserResult.Error.DuplicateUsername
                } else {
                    CreateUserResult.Error.InvalidInput(response.message)
                }
            }
            ErrorCodes.INVALID_INPUT, ErrorCodes.VALIDATION_FAILED -> {
                CreateUserResult.Error.InvalidInput(response.message)
            }
            else -> CreateUserResult.Error.ServerError(response.message, response.errorCode)
        }
    }
}
